src/oidcendpoint/oauth2/introspection.py

```diff
@@ -23,59 +22,37 @@ class Introspection(Endpoint):
     def __init__(self, **kwargs):
         Endpoint.__init__(self, **kwargs)
         self.offset = kwargs.get("offset", 0)
-        self.enable_claims_per_client = kwargs.get("enable_claims_per_client", False)
-
-    def get_client_id_from_token(self, endpoint_context, token, request=None):
-        """
-        Will try to match tokens against information in the session DB.
-
-        :param endpoint_context:
-        :param token:
-        :param request:
-        :return: client_id if there was a match
-        """
-        sinfo = endpoint_context.sdb[token]
-        return sinfo["authn_req"]["client_id"]
-
-    def _get_client_claims(self, token):
-        client_id = self.get_client_id_from_token(self.endpoint_context, token)
-        client = self.endpoint_context.cdb.get(client_id, {})
-        return client.get("introspection_claims")
-
-    def _get_user_info(self, token_info):
-        user_id = self.endpoint_context.sdb.sso_db.get_uid_by_sid(token_info["sid"])
-        return self.endpoint_context.userinfo(user_id, client_id=None)
-
-    def _add_claims(self, token_info, claims, payload):
-        user_info = self._get_user_info(token_info)
-        for attr in claims:
-            try:
-                payload[attr] = user_info[attr]
-            except KeyError:
-                pass
-
-    def _introspect(self, token):
-        try:
-            info = self.endpoint_context.sdb[token]
-        except (KeyError, UnknownToken):
-            return None
 
+    def _introspect(self, token, session_info):
         # Make sure that the token is an access_token or a refresh_token
-        if token != info.get("access_token") and token != info.get("refresh_token"):
+        if token.type not in ["access_token", "refresh_token"]:
             return None
 
-        eat = info.get("expires_at")
-        if eat and eat < utc_time_sans_frac():
+        if not token.is_active():
             return None
 
-        if info:  # Now what can be returned ?
-            ret = info.to_dict()
-            ret["iss"] = self.endpoint_context.issuer
-
-            if "scope" not in ret:
-                ret["scope"] = " ".join(info["authn_req"]["scope"])
-
-            return ret
+        scope = token.scope
+        if not scope:
+            scope = session_info["grant"].scope
+        aud = token.resources
+        if not aud:
+            aud = session_info["grant"].resources
+
+        _csi = session_info["client_session_info"]
+        ret = {
+            "active": True,
+            "scope": " ".join(scope),
+            "client_id": session_info["client_id"],
+            "token_type": token.type,
```

Comment:
`token_type` is not supposed to be `access_token` or `refresh_token`. It should be `bearer` or `mac` (https://tools.ietf.org/html/rfc6749#section-7.1)

Reply:
OK! Presently we're only dealing with `bearer` type tokens.